### ISSUE TITLE
torch.compile Introduces Significant Numerical Errors in Boltzmann Machine Model

### ISSUE THREAD

**Issue by zhouxiaoyaozzz** [pytorch/pytorch#162723]
### 🐛 Describe the bug

Description
When using torch.compile() on a custom Boltzmann Machine model, the compiled version produces significantly different numerical results compared to the original model. The compilation process introduces substantial errors in hidden layer activations, energy calculations, and sampling behavior, making the compiled model functionally incorrect.
code:
'''
import torch
import torch.nn as nn
class BoltzmannMachine(nn.Module):
    def __init__(self, num_visible, num_hidden, learning_rate=0.01, k=1):
        super(BoltzmannMachine, self).__init__()
        self.num_visible = num_visible
        self.num_hidden = num_hidden
        self.k = k
        self.learning_rate = learning_rate
        self.weights = nn.Parameter(torch.randn(num_visible, num_hidden) * 0.01)
        self.visible_bias = nn.Parameter(torch.zeros(num_visible))
        self.hidden_bias = nn.Parameter(torch.zeros(num_hidden))
    def sample_hidden(self, visible_prob):
        hidden_activations = torch.matmul(visible_prob, self.weights) + self.hidden_bias
        hidden_prob = torch.sigmoid(hidden_activations)
        return torch.bernoulli(hidden_prob), hidden_prob
    def sample_visible(self, hidden_prob):
        visible_activations = torch.matmul(hidden_prob, self.weights.t()) + self.visible_bias
        visible_prob = torch.sigmoid(visible_activations)
        return torch.bernoulli(visible_prob), visible_prob
    def contrastive_divergence(self, visible_data):
        batch_size = visible_data.shape[0]
        pos_hidden_prob = torch.sigmoid(torch.matmul(visible_data, self.weights) + self.hidden_bias)
        pos_hidden_sample = torch.bernoulli(pos_hidden_prob)
        pos_associations = torch.matmul(visible_data.t(), pos_hidden_prob)
        neg_visible = visible_data.clone()
        for _ in range(self.k):
            neg_hidden_sample, neg_hidden_prob = self.sample_hidden(neg_visible)
            neg_visible_sample, neg_visible_prob = self.sample_visible(neg_hidden_sample)
            neg_visible = neg_visible_prob
        neg_associations = torch.matmul(neg_visible.t(), neg_hidden_prob)
        self.weights.data += self.learning_rate * (pos_associations - neg_associations) / batch_size
        self.visible_bias.data += self.learning_rate * torch.mean(visible_data - neg_visible, dim=0)
        self.hidden_bias.data += self.learning_rate * torch.mean(pos_hidden_prob - neg_hidden_prob, dim=0)
        error = torch.mean((visible_data - neg_visible) ** 2)
        return error.item()
    def free_energy(self, v):
        vbias_term = torch.matmul(v, self.visible_bias)
        wx_b = torch.matmul(v, self.weights) + self.hidden_bias
        hidden_term = torch.sum(torch.log(1 + torch.exp(wx_b)), dim=1)
        return -hidden_term - vbias_term
    def forward(self, visible_data, n_gibbs=100):
        samples = visible_data.clone()
        for _ in range(n_gibbs):
            _, hidden_prob = self.sample_hidden(samples)
            samples, _ = self.sample_visible(hidden_prob)
        return samples

if __name__ == "__main__":
    torch.manual_seed(42)
    torch.cuda.manual_seed_all(42)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

    num_visible = 784
    num_hidden = 128
    batch_size = 8
    learning_rate = 0.01

    input_tensor = torch.randint(0, 2, (batch_size, num_visible), dtype=torch.float32)
    initial_state = torch.randint(0, 2, (5, num_visible), dtype=torch.float32)

    model = BoltzmannMachine(num_visible, num_hidden, learning_rate)
    for param in model.parameters():
        param.requires_grad = False

    model.eval()
    with torch.no_grad():
        samples_normal = model(initial_state)
        energy_normal = model.free_energy(input_tensor)
        recon_error_normal = model.contrastive_divergence(input_tensor)
        _, hidden_prob_normal = model.sample_hidden(input_tensor)
        _, visible_prob_normal = model.sample_visible(hidden_prob_normal)

    compiled_model = torch.compile(model)
    compiled_model.eval()
    with torch.no_grad():
        samples_compiled = compiled_model(initial_state)
        energy_compiled = compiled_model.free_energy(input_tensor)
        recon_error_compiled = compiled_model.contrastive_divergence(input_tensor)
        _, hidden_prob_compiled = compiled_model.sample_hidden(input_tensor)
        _, visible_prob_compiled = compiled_model.sample_visible(hidden_prob_compiled)

    diff_hidden = torch.abs(hidden_prob_normal - hidden_prob_compiled)
    diff_visible = torch.abs(visible_prob_normal - visible_prob_compiled)
    diff_energy = torch.abs(energy_normal - energy_compiled)
    diff_samples = torch.abs(samples_normal.float() - samples_compiled.float())  # 转换为float

    print(f"\n● Output Shape Verification:")
    print(f"  Original samples shape: {samples_normal.shape}")
    print(f"  Compiled samples shape: {samples_compiled.shape}")
    print(f"  Shape consistency: {samples_normal.shape == samples_compiled.shape}")

    print(f"\n● Numerical Consistency Check (rtol=0.01, atol=0.001):")
    max_abs_diff_hidden = torch.max(diff_hidden).item()
    max_rel_diff_hidden = torch.max(diff_hidden / (torch.abs(hidden_prob_normal) + 1e-8)).item()
    mismatch_count_hidden = torch.sum(diff_hidden > 0.001).item()
    total_elements_hidden = hidden_prob_normal.numel()

    print(
        f"  Hidden layer - Mismatched elements: {mismatch_count_hidden} / {total_elements_hidden} ({mismatch_count_hidden / total_elements_hidden * 100:.1f}%)")
    print(f"  Hidden layer - Max absolute difference: {max_abs_diff_hidden:.6f}")
    print(f"  Hidden layer - Max relative difference: {max_rel_diff_hidden:.6f}")

    atol_violation_hidden = max_abs_diff_hidden > 0.001
    rtol_violation_hidden = max_rel_diff_hidden > 0.01

    if not atol_violation_hidden and not rtol_violation_hidden:
        print(f"  ✅ Hidden layer passed consistency check")
    else:
        print(f"  ❌ Hidden layer failed consistency check")
        if atol_violation_hidden:
            print(f"    Absolute tolerance violation: {max_abs_diff_hidden:.6f} > 0.001")
        if rtol_violation_hidden:
            print(f"    Relative tolerance violation: {max_rel_diff_hidden:.6f} > 0.01")

    print(f"\n● Probability Distribution Differences:")
    prob_max_diff_hidden = torch.max(diff_hidden).item()
    prob_mean_diff_hidden = torch.mean(diff_hidden).item()
    prob_l2_diff_hidden = torch.norm(diff_hidden).item()

    prob_max_diff_visible = torch.max(diff_visible).item()
    prob_mean_diff_visible = torch.mean(diff_visible).item()
    prob_l2_diff_visible = torch.norm(diff_visible).item()

    print(f"  Hidden layer - Max probability difference: {prob_max_diff_hidden:.8f}")
    print(f"  Hidden layer - Mean probability difference: {prob_mean_diff_hidden:.8f}")
    print(f"  Hidden layer - L2 norm difference: {prob_l2_diff_hidden:.8f}")

    print(f"  Visible layer - Max probability difference: {prob_max_diff_visible:.8f}")
    print(f"  Visible layer - Mean probability difference: {prob_mean_diff_visible:.8f}")
    print(f"  Visible layer - L2 norm difference: {prob_l2_diff_visible:.8f}")

    print(f"\n● Sample Consistency Check:")

    sample_diff_count = torch.sum(samples_normal != samples_compiled).item()
    sample_total = samples_normal.numel()
    sample_agreement = (sample_total - sample_diff_count) / sample_total * 100
    print(f"  Sample differences: {sample_diff_count} / {sample_total}")
    print(f"  Sample agreement rate: {sample_agreement:.2f}%")
    print(f"  Complete sample consistency: {sample_diff_count == 0}")

    print(f"\n● Energy Value Comparison:")
    energy_diff = torch.mean(diff_energy).item()
    energy_mean_normal = torch.mean(energy_normal).item()
    energy_mean_compiled = torch.mean(energy_compiled).item()
    print(f"  Original energy mean: {energy_mean_normal:.6f}")
    print(f"  Compiled energy mean: {energy_mean_compiled:.6f}")
    print(f"  Energy difference: {energy_diff:.6f}")
    print(f"  Energy consistency: {energy_diff < 0.01}")

    print(f"\n● Reconstruction Error Comparison:")
    recon_diff = abs(recon_error_normal - recon_error_compiled)
    print(f"  Original reconstruction error: {recon_error_normal:.6f}")
    print(f"  Compiled reconstruction error: {recon_error_compiled:.6f}")
    print(f"  Reconstruction error difference: {recon_diff:.6f}")
    print(f"  Reconstruction consistency: {recon_diff < 0.01}")

    print(f"\n● Sample Value Analysis:")
    print(f"  Original samples - True count: {torch.sum(samples_normal).item()}/{sample_total}")
    print(f"  Compiled samples - True count: {torch.sum(samples_compiled).item()}/{sample_total}")
    print(f"  Sample value agreement: {sample_agreement:.2f}%")
'''
output:
'''
● Output Shape Verification:
  Original samples shape: torch.Size([5, 784])
  Compiled samples shape: torch.Size([5, 784])
  Shape consistency: True

● Numerical Consistency Check (rtol=0.01, atol=0.001):
  Hidden layer - Mismatched elements: 1024 / 1024 (100.0%)
  Hidden layer - Max absolute difference: 0.033240
  Hidden layer - Max relative difference: 0.057393
  ❌ Hidden layer failed consistency check
    Absolute tolerance violation: 0.033240 > 0.001
    Relative tolerance violation: 0.057393 > 0.01

● Probability Distribution Differences:
  Hidden layer - Max probability difference: 0.03323960
  Hidden layer - Mean probability difference: 0.02678068
  Hidden layer - L2 norm difference: 0.86185277
  Visible layer - Max probability difference: 0.04800451
  Visible layer - Mean probability difference: 0.01256461
  Visible layer - L2 norm difference: 1.28579807

● Sample Consistency Check:
  Sample differences: 1925 / 3920
  Sample agreement rate: 50.89%
  Complete sample consistency: False

● Energy Value Comparison:
  Original energy mean: -89.674187
  Compiled energy mean: -97.408691
  Energy difference: 7.734510
  Energy consistency: False

● Reconstruction Error Comparison:
  Original reconstruction error: 0.249770
  Compiled reconstruction error: 0.244479
  Reconstruction error difference: 0.005291
  Reconstruction consistency: True

● Sample Value Analysis:
  Original samples - True count: 1946.0/3920
  Compiled samples - True count: 1949/3920
  Sample value agreement: 50.89%
'''

### Versions

Collecting environment information...
PyTorch version: 2.8.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.5.0-6ubuntu2) 9.5.0
Clang version: Could not collect
CMake version: version 4.0.0
Libc version: glibc-2.39

Python version: 3.9.7 (default, Jul 16 2025, 16:34:47) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-28-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060 Laptop GPU
Nvidia driver version: 575.64.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
架构： x86_64
CPU 运行模式： 32-bit, 64-bit
Address sizes: 39 bits physical, 48 bits virtual
字节序： Little Endian
CPU: 32
在线 CPU 列表： 0-31
厂商 ID： GenuineIntel
型号名称： Intel(R) Core(TM) i9-14900HX
CPU 系列： 6
型号： 183
每个核的线程数： 2
每个座的核数： 24
座： 1
步进： 1
CPU(s) scaling MHz: 23%
CPU 最大 MHz： 5800.0000
CPU 最小 MHz： 800.0000
BogoMIPS： 4838.40
标记： fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
虚拟化： VT-x
L1d 缓存： 896 KiB (24 instances)
L1i 缓存： 1.3 MiB (24 instances)
L2 缓存： 32 MiB (12 instances)
L3 缓存： 36 MiB (1 instance)
NUMA 节点： 1
NUMA 节点0 CPU： 0-31
Vulnerability Gather data sampling: Not affected
Vulnerability Ghostwrite: Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] intel-cmplr-lib-ur==2025.1.0
[pip3] intel-openmp==2025.1.0
[pip3] mkl==2025.1.0
[pip3] mkl-include==2025.1.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.18.0
[pip3] onnxruntime==1.19.2
[pip3] optree==0.17.0
[pip3] tbb==2022.1.0
[pip3] tcmlib==1.3.0
[pip3] torch==2.8.0
[pip3] torchdiffeq==0.2.5
[pip3] torchvision==0.23.0
[pip3] triton==3.4.0
[pip3] umf==0.10.0
[conda] Could not collect

cc @chauhang @penguinwu

**Comment by jansel** [comment]
This report also looks invalid for a similar reason as #162722 

Your repro script is flawed, if I remove `torch.compile`:
```patch
diff --git a/repro_issue_162723.py b/repro_issue_162723.py
index 8c09f70aefb..73e50a269c2 100644
--- a/repro_issue_162723.py
+++ b/repro_issue_162723.py
@@ -73,7 +73,7 @@ if __name__ == "__main__":
         _, hidden_prob_normal = model.sample_hidden(input_tensor)
         _, visible_prob_normal = model.sample_visible(hidden_prob_normal)
 
-    compiled_model = torch.compile(model)
+    compiled_model = model
     compiled_model.eval()
     with torch.no_grad():
         samples_compiled = compiled_model(initial_state)
```

I get the output:
```
● Output Shape Verification:
  Original samples shape: torch.Size([5, 784])
  Compiled samples shape: torch.Size([5, 784])
  Shape consistency: True

● Numerical Consistency Check (rtol=0.01, atol=0.001):
  Hidden layer - Mismatched elements: 1024 / 1024 (100.0%)
  Hidden layer - Max absolute difference: 0.033008
  Hidden layer - Max relative difference: 0.057356
  ❌ Hidden layer failed consistency check
    Absolute tolerance violation: 0.033008 > 0.001
    Relative tolerance violation: 0.057356 > 0.01

● Probability Distribution Differences:
  Hidden layer - Max probability difference: 0.03300810
  Hidden layer - Mean probability difference: 0.02677627
  Hidden layer - L2 norm difference: 0.86165351
  Visible layer - Max probability difference: 0.04800844
  Visible layer - Mean probability difference: 0.01250973
  Visible layer - L2 norm difference: 1.28038824

● Sample Consistency Check:
  Sample differences: 1992 / 3920
  Sample agreement rate: 49.18%
  Complete sample consistency: False

● Energy Value Comparison:
  Original energy mean: -89.674187
  Compiled energy mean: -97.408691
  Energy difference: 7.734512
  Energy consistency: False

● Reconstruction Error Comparison:
  Original reconstruction error: 0.249770
  Compiled reconstruction error: 0.244665
  Reconstruction error difference: 0.005106
  Reconstruction consistency: True

● Sample Value Analysis:
  Original samples - True count: 1946.0/3920
  Compiled samples - True count: 1958.0/3920
  Sample value agreement: 49.18%
```
which is almost the same as with `torch.compile`, so I think this methodology is incorrect.

I suspect this is due to random number generation ops in the model similar to the other invalid report at #162722 -- but I have not debugged it further than the above.

@zhouxiaoyaozzz please run this sanity check any other issues you have opened and close them if they fail it.
